client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -1,182 +1,136 @@
+    */
+  def borrow(key: RequestKey): Task[NextConnection] = connectionBuffer.flatMap{q =>
+    q.full.get.flatMap { full =>
+      if (full) {
+        q.dequeue1.flatMap { conn =>
+          conn.isClosed.flatMap { closed =>
+            if (!closed && key == conn.requestKey) {
+              Task.delay(NextConnection(conn, false))
+            }
+            else if (!closed) {
+              q.enqueue1(conn) >>  borrow(key)
+            }
+            else {
+              q.build(key) >> borrow(key)
+            }
           }
-          else logger.debug(s"Connection could not be recycled, no pending requests. Shrinking pool: ${stats}")
         }
       }
-      else if (!connection.isClosed) {
-        logger.debug(s"Shutting down connection after pool closure: ${stats}")
-        connection.shutdown()
-        allocated -= 1
+      else {
+        q.build(key) >> borrow(key)
       }
     }
   }
 
-  override def invalidate(connection: A): Task[Unit] =
-    Task.delay(disposeConnection(connection.requestKey, Some(connection)))
-
-  private def disposeConnection(key: RequestKey, connection: Option[A]) = {
-    logger.debug(s"Disposing of connection: ${stats}")
-    synchronized {
-      allocated -= 1
-      connection.foreach { s => if (!s.isClosed) s.shutdown() }
+  /**
+    * This checks to see if the connection is already closed. If it is, then we are finished.
+    * If it is recyclable it is added back to the connection buffer,
+    * and if it is not it is shutdown as we no longer can use the connection.
+    * @param connection The Connection We Are Letting Go Of
+    * @return A Task of Unit
+    */
+  def release(connection: A): Task[Unit] =  connectionBuffer.flatMap{ q =>
+    connection.isClosed.flatMap{ closed =>
+      connection.isRecyclable.flatMap{ recyclable =>
+        if (closed){
+          Task.now(())
+        } else if (recyclable){
+          q.enqueue1(connection)
+        } else {
+          connection.shutdown()
+        }
+      }
     }
   }
 
-  def shutdown() = Task.delay {
-    logger.info(s"Shutting down connection pool: ${stats}")
-    synchronized {
-      if (!isClosed) {
-        isClosed = true
-        idleQueue.foreach(_.shutdown())
-        allocated = 0
-      }
-    }
+  /**
+    * We invalidate connections, as leases of connections have been given out
+    * @param connection The connection to invalidate
+    * @return A Task of Unit
+    */
+  override def invalidate(connection: A): Task[Unit] =
+    connection.isClosed.flatMap{ closed => if (!closed) connection.shutdown() else Task.now(()) }
+
+  /**
+    * Transforms the Queue of Connections into A Task of Unit by evaluating an invalidation of all
+    * the connections. Emptying out the Pool completely when the effect is evaluated.
+    * @return
+    */
+  def shutdown() : Task[Unit] = connectionBuffer.flatMap{ q =>
+    q.dequeue.evalMap[Task, Task, Unit]{invalidate}.run
```

Comment:
I believe this is run in parallel. invalidate is a description of a computation that is run by the Stream, so each computation is triggered as it hits that element of the stream, and the Strategy is where it completes asynchronously. I think...